cmd/acra-translator/server/server.go

```diff
@@ -193,129 +187,216 @@ func (server *ReaderServer) HandleConnectionString(parentContext context.Context
+		server.startHTTP(parentContext, logger, decryptorData, errCh, listener)
+	}
 
-func withHandlerName(ctx context.Context, name string) context.Context {
-	return context.WithValue(ctx, handlerName{}, name)
-}
+	// provide way to register new services and custom server
+	if server.config.IncomingConnectionGRPCString() != "" {
+		listener, err := network.Listen(server.config.IncomingConnectionGRPCString())
+		if err != nil {
+			log.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantAcceptNewHTTPConnection).
+				Errorln("Can't create HTTP listener from specified connection string")
+			return
+		}
+		server.listenerGRPC = listener
+		server.startGrpc(logger, decryptorData, errCh, listener)
+	}
 
-func getHandlerName(ctx context.Context) string {
-	if s, ok := ctx.Value(handlerName{}).(string); ok {
-		return s
+	select {
+	case <-parentContext.Done():
+		break
+	case <-errCh:
+		// error that occurred in background goroutines will be logged, so just break and exit with timeout,
+		// according to defer func
+		break
 	}
-	return "undefined"
+	return
 }
 
-// Start setups gRPC handler or HTTP handler, poison records callbacks and starts listening to connections.
-func (server *ReaderServer) Start(parentContext context.Context) {
-	logger := logging.GetLoggerFromContext(parentContext)
-	poisonCallbacks := base.NewPoisonCallbackStorage()
+func (server *ReaderServer) startHTTP(parentContext context.Context, logger *log.Entry, decryptorData *common.TranslatorData, errCh chan<- error, listener net.Listener) {
+	server.backgroundWorkersSync.Add(1)
+	go func() {
+		defer server.backgroundWorkersSync.Done()
+		httpContext := logging.SetLoggerToContext(parentContext, logger.WithField(ConnectionTypeKey, HTTPConnectionType))
+		httpDecryptor, err := http_api.NewHTTPConnectionsDecryptor(decryptorData)
+		logger.WithField("connection_string", server.config.IncomingConnectionHTTPString()).Infof("Start process HTTP requests")
+		if err != nil {
+			log.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantHandleHTTPConnection).
+				Errorln("Can't create HTTP decryptor")
+			errCh <- err
+			return
+		}
+		server.httpDecryptor = httpDecryptor
+		err = server.HandleHTTPConnection(httpContext, listener, server.config.IncomingConnectionHTTPString(), server.processHTTPConnection)
+		if err != nil {
+			// It is a "normal" case, when we shutdown service - we call 'cancel' on main level in SIGTERM/SIGINT handlers,
+			// so let's avoid treating context.Canceled as error here
+			if !errors.As(err, &context.Canceled) {
+				log.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantHandleHTTPConnection).
+					Errorln("Took error on handling HTTP requests")
+				server.Stop()
+				errCh <- err
+			}
+			return
+		}
+	}()
+}
+
+func (server *ReaderServer) startGrpc(logger *log.Entry, decryptorData *common.TranslatorData, errCh chan<- error, listener net.Listener) {
```

Comment:
```suggestion
func (server *ReaderServer) startGRPC(logger *log.Entry, decryptorData *common.TranslatorData, errCh chan<- error, listener net.Listener) {
```

I see that functions and variables in code use HTTP and GRPC as ALL CAPS format, let's continue doing so.

Reply:
ok